rfc/bulksave.md

```diff
@@ -0,0 +1,75 @@
+# To improve the spinnaker sql storage performance by using bulk save api
+
+| | |
+|-|-|
+| **Status**     | _**Proposed**, Accepted, Implemented, Obsolete_ |
+| **RFC #**      | |
+| **Author(s)**  | Sanjeev Thatiparthi (@sanopsmx) (https://github.com/sanopsmx)
+| **SIG / WG**   | sig-security
+
+## Overview
+
+https://github.com/spinnaker/spinnaker/issues/6147
+
+Spinnaker API currently supports saving one pipeline at a time. In order to save multiple pipelines,
+the external system needs to iteratively call the save pipeline method. In a good deployment,
+the pipeline save goes through multiple hops and can take 2 to 5s and can be slower in other deployments.
+The overall time to update all the pipelines can run into hours.
+
+### Goals and Non-Goals
+
+To improve the performance of gate, orca and front50 microservices by using bulk insert query.
+We need to write a bulk save api which accepts an list of pipelines in a single rest api call
+and save them in front50 using bulk insert query.
+
+## Motivation and Rationale
+
+In many deployments, the pipelines are managed with templates and pipeline json structures are generated
+from these templates based on enterprise requirements and then saved to Spinnaker using the API.
+This operation occurs frequently enough that there is a requirement for efficiency in saving bulk pipelines in Spinnaker.
+The changes to pipelines can occur in a Spinnaker deployments because of
+
+1. Changes in policies associated with pipelines
+2. Changes to metadata in the pipelines for better analytics of executed pipelines
+3. Synchronizing pipelines in Spinnaker with Git spanning multiple applications
+
+Applying the changes to pipelines with one pipeline at a time with currently available API takes hours to complete.
+Since the operation of bulk updates is performed often and multiple organizations,
+it makes sense to have Spinnaker support batch updates of pipelines.
```

Comment:
Could you go into detail about what other approaches you've tried that lead to your need for a Bulk Save API? You mention that you're currently calling the API sequentially but have you tried calling the API concurrently? If so, what were the results? Have you tried using the `savePipeline` task in Orca and submitting multiple tasks? I'm interested to hear about any performance issues you've experienced using these approaches if you have. If not, why? 

I believe I asked @emagana to include some of this information in the original design proposal.

Reply:
To add some helpful advice for future proposals, including data like that is critical for the community, not just the TOC, to evaluate the need for features, especially performance improvements. Understanding the full scope of what's been tried will enable us to either accept the proposal as is or offer alternative approaches and iterate on a design that suits the entire community. To be clear, I don't believe this is one of those cases. Just offering some advice on writing effective proposals :D.

Reply:
https://github.com/spinnaker/governance/pull/194

This has the history of the previous implementation.

@ajordens @robzienert 
has suggested this approach.